src/appleseed/renderer/kernel/lighting/lighttree.cpp

```diff
@@ -0,0 +1,332 @@
+
+                    file.draw_aabb(bbox_left, color, Width);
+                    file.draw_aabb(bbox_right, color, Width);
+                }
+            }
+        }
+    }
+    else
+    {
+        const std::string filename = filename_base + ".py";
+        foundation::VPythonFile file(filename.c_str());
+        file.draw_axes(Width);
+        const char* color = "color.yellow";
+
+        // Draw the initial bbox.
+        file.draw_aabb(root_bbox, color, Width);
+
+        // Find the parent node to draw child bboxes from.
+        for (size_t i = 0; i < m_nodes.size(); i++)
+        {
+            if (m_nodes[i].is_leaf())
+                continue;
+
+            // Make even levels red and odd green
+            color = m_nodes[i].get_level() % 2 ? "color.red"
+                                               : "color.green";
+
+            const auto& bbox_left = m_nodes[i].get_left_bbox();
+            const auto& bbox_right = m_nodes[i].get_right_bbox();
+
+            file.draw_aabb(bbox_left, color, Width);
+            file.draw_aabb(bbox_right, color, Width);
+        }
+    }
+}
+
+//
+// Calculate the tree depth and assign total luminance to each node of the tree.
+// Total luminance represents the sum of all its child nodes luminances.
+//
+
+float LightTree::recursive_node_update(size_t node_index, size_t node_level)
+{
+    float luminance = 0.0f;
+
+    if (!m_nodes[node_index].is_leaf())
+    {
+        const auto& child1 = m_nodes[node_index].get_child_node_index();
+        const auto& child2 = m_nodes[node_index].get_child_node_index() + 1;
+
+        float luminance1 = recursive_node_update(child1, node_level + 1);
+        float luminance2 = recursive_node_update(child2, node_level + 1);
+
+        luminance = luminance1 + luminance2;
+    }
+    else
+    {
+        // Access the light intensity value.
+        size_t item_index = m_nodes[node_index].get_item_index();
+        size_t light_source_index = m_items[item_index].m_light_index;
+        Spectrum spectrum = m_light_sources[light_source_index]->get_intensity();
+
+        // Luminance is the average accross all the spectrum channels.
+        for (size_t i = 0; i < spectrum.size(); i++)
+        {
+            luminance += spectrum[i];
+        }
+        luminance /= spectrum.size();
+
+        // Modify the tree depth if the branch is deeper than the other branches
+        // visited so far.
+        if (m_tree_depth < node_level)
+            m_tree_depth = node_level;
+    }
+
+    m_nodes[node_index].set_luminance(luminance);
+    m_nodes[node_index].set_level(node_level);
+
+    return luminance;
+}
+
+std::pair<size_t, float> LightTree::sample(
+    const foundation::Vector3d&     surface_point,
+    float                           s) const
+{
+    float light_probability = 1.0;
+    size_t node_index = 0;
+
+    while (!m_nodes[node_index].is_leaf())
+    {
+        // LightTreeNode.
```

Comment:
I use `LightTreeNode` type and I left it here as a hint for the auto types because it might not be intuitive for someone who is going through the code for the first time and needs to change something.
If you think I should remove it, I will.